src/plot.js

```diff
@@ -96,13 +125,69 @@ export function plot(options = {}) {
-    if (node != null) svg.appendChild(node);
+  // When faceting, render axes for fx and fy instead of x and y.
+  const {fx, fy} = scales;
+  const axisY = axes[facets !== undefined && fy ? "fy" : "y"];
+  const axisX = axes[facets !== undefined && fx ? "fx" : "x"];
+  if (axisY) svg.appendChild(axisY.render(null, scales, dimensions));
+  if (axisX) svg.appendChild(axisX.render(null, scales, dimensions));
+
+  // Render (possibly faceted) marks.
+  if (facets !== undefined) {
+    const fyDomain = fy && fy.domain();
+    const fxDomain = fx && fx.domain();
+    const fyMargins = fy && {marginTop: 0, marginBottom: 0, height: fy.bandwidth()};
+    const fxMargins = fx && {marginRight: 0, marginLeft: 0, width: fx.bandwidth()};
+    const subdimensions = {...dimensions, ...fxMargins, ...fyMargins};
+    const indexByFacet = facetMap(facetChannels);
+    facets.forEach(([key], i) => indexByFacet.set(key, i));
+    select(svg)
+        .call(g => {
+          if (fy && axes.y) {
+            const axis1 = axes.y, axis2 = nolabel(axis1);
+            const j = axis1.labelAnchor === "bottom" ? fyDomain.length - 1 : axis1.labelAnchor === "center" ? fyDomain.length >> 1 : 0;
+            g.selectAll()
+              .data(fyDomain)
+              .enter()
+              .append((ky, i) => (i === j ? axis1 : axis2).render(
+                fx && where(fxDomain, kx => indexByFacet.has([kx, ky])),
+                scales,
+                {...dimensions, ...fyMargins, offsetTop: fy(ky)}
+              ));
+          }
+          if (fx && axes.x) {
+            const axis1 = axes.x, axis2 = nolabel(axis1);
+            const j = axis1.labelAnchor === "right" ? fxDomain.length - 1 : axis1.labelAnchor === "center" ? fxDomain.length >> 1 : 0;
+            const {marginLeft, marginRight} = dimensions;
+            g.selectAll()
+              .data(fxDomain)
+              .enter()
+              .append((kx, i) => (i === j ? axis1 : axis2).render(
+                fy && where(fyDomain, ky => indexByFacet.has([kx, ky])),
+                scales,
+                {...dimensions, ...fxMargins, labelMarginLeft: marginLeft, labelMarginRight: marginRight, offsetLeft: fx(kx)}
+              ));
+          }
+        })
+        .call(g => g.selectAll()
+          .data(facetKeys(scales).filter(indexByFacet.has, indexByFacet))
+          .enter()
+          .append("g")
+            .attr("transform", facetTranslate(fx, fy))
+            .each(function(key) {
+              const j = indexByFacet.get(key);
+              for (const [mark, {channels, values, index, faceted}] of stateByMark) {
+                const renderIndex = mark.filter(faceted ? index[j] : index, channels, values);
+                const node = mark.render(renderIndex, scales, values, subdimensions);
+                if (node != null) this.appendChild(node);
+              }
+            }));
+  } else {
+    for (const [mark, {channels, values, index}] of stateByMark) {
```

Comment:
For code readability also, it would be nice to have the simpler case (non-faceted) come first, so that the reader understands what this part is doing before diving in (or skipping) the faceted part.

Reply:
That’s appealing, but it also has the consequence of further separating these two similar loops, and I think it’s nice if they’re close to each other so it’s more obvious the logic must be keep consistent.

https://github.com/observablehq/plot/blob/6a2c559a22c7164c387d0d9cae6de5824e5b2857/src/plot.js#L179-L183

https://github.com/observablehq/plot/blob/6a2c559a22c7164c387d0d9cae6de5824e5b2857/src/plot.js#L186-L190

That’s because in the faceting case there is a bunch of additional work to render the facet axes.

Reply:
yes—it's fine like that